sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -128,90 +127,65 @@ package object expressions  {
-      var matches: (Seq[Attribute], Seq[String]) = nameParts match {
-        case dbPart +: tblPart +: name +: nestedFields =>
-          val key = (dbPart.toLowerCase(Locale.ROOT),
-            tblPart.toLowerCase(Locale.ROOT), name.toLowerCase(Locale.ROOT))
-          val attributes = collectMatches(name, qualified3Part.get(key)).filter {
-            a => (resolver(dbPart, a.qualifier.head) && resolver(tblPart, a.qualifier.last))
-          }
-          (attributes, nestedFields)
-        case _ =>
-          (Seq.empty, Seq.empty)
+      // Collect attributes that match the given name and qualifier.
+      // A match occurs if
+      //   1) the given name matches the attribute's name according to the resolver.
+      //   2) the given qualifier is a subset of the attribute's qualifier.
+      def collectMatches(
+          name: String,
+          qualifier: Seq[String],
+          candidates: Option[Seq[Attribute]]): Seq[Attribute] = {
+        candidates.toSeq.flatMap(_.collect {
+          case a if resolver(name, a.name) && matchQualifier(qualifier, a.qualifier) =>
+            a.withName(name)
+        })
       }
 
-      // If there are no matches, then find matches for the given name assuming that
-      // the 1st part is a qualifier (i.e. table name, alias, or subquery alias) and the
-      // 2nd part is the actual name. This returns a tuple of
-      // matched attributes and a list of parts that are to be resolved.
-      //
-      // For example, consider an example where "a" is the table name, "b" is the column name,
-      // and "c" is the struct field name, i.e. "a.b.c". In this case, Attribute will be "a.b",
-      // and the second element will be List("c").
-      if (matches._1.isEmpty) {
-        matches = nameParts match {
-          case qualifier +: name +: nestedFields =>
-            val key = (qualifier.toLowerCase(Locale.ROOT), name.toLowerCase(Locale.ROOT))
-            val attributes = collectMatches(name, qualified.get(key)).filter { a =>
-              resolver(qualifier, a.qualifier.last)
-            }
-            (attributes, nestedFields)
-          case _ =>
-            (Seq.empty[Attribute], Seq.empty[String])
+      // Iterate each string in `nameParts` in a reverse order and try to match the attributes
+      // considering the current string as the attribute name. For example, if `nameParts` is
+      // Seq("a", "b", "c"), the match will be performed in the following order:
+      // 1) name = "c", qualifier = Seq("a", "b")
+      // 2) name = "b", qualifier = Seq("a")
+      // 3) name = "a", qualifier = Seq()
+      // Note that the match is performed in the reverse order in order to match the longest
+      // qualifier as possible. If a match is found, the remaining portion of `nameParts`
+      // is also returned as nested fields.
+      val matches = nameParts.zipWithIndex.reverseIterator.flatMap { case (name, index) =>
```

Comment:
I'd suggest writing `while` loop manually, so that we can stop the loop once a match is found. e.g.
```
var candidates: Seq[Attribute] = Nil
var nestedFields = nameParts
var i = 0
while (i < nameParts.length && candidates.isEmpty) {
  ...
}
```

Reply:
The type of `matches` here is `Iterator[(Seq[Attribute], Seq[String])]`.  And since I am doing `val (candidates, nestedFields) = matches.next` below, it will find only the first match.

If this was not clear, I can make this to an explicit loop. Please let me know, thanks!

Reply:
Yea I know we only get the first match, I mean we collect second, third ... matches unnecessarily. Since we only need the first match, it's clearer to stop looking for matches if one match is found.

Reply:
Oh, since we are doing `(matches.hasNext) { matches.next }`, only the first match will be collected. But I will do an explicit loop to make the intention clear. Thanks!